### ISSUE TITLE
authenticateUser results in Unable to verify secret hash or client 

### ISSUE THREAD

**Issue by kbokarius** [furaiev/amazon-cognito-identity-dart-2#32]
```
class Cognito {
  static const _userPoolId = 'xxx';
  static const _clientId = 'xxx';
  static const _clientSecret = 'xxx';

  final CognitoUserPool userPool = new CognitoUserPool(_userPoolId, _clientId, clientSecret: _clientSecret);

  static final Cognito _instance = Cognito._internal();

  factory Cognito() {
    return _instance;
  }

  Cognito._internal();

  Future signUpAndGetToken(String email, String password) async {
    await userPool.signUp(email, password);
    return await signInAndGetToken(email, password);
  }

  Future signInAndGetToken(String email, String password) async {
    final cognitoUser = new CognitoUser(email, new Cognito().userPool, clientSecret: _clientSecret);
    final authDetails = new AuthenticationDetails(
      username: email,
      password: password,
    );
    CognitoUserSession session = await cognitoUser.authenticateUser(authDetails);
    return session.getAccessToken().getJwtToken();
  }
```

I'm able to sign up a new user without an error, but the `CognitoUserSession session = await cognitoUser.authenticateUser(authDetails);` results in `name: NotAuthorizedException, message: Unable to verify secret hash for client`

I'm passing the same client secret in both the call to signUp and authenticateUser, so why would signUp be working but not authenticateUser?

**Comment by droidkfx** [comment]
If you are server or need admin (programmatic user access without an actual user) access to Cognito ignore this comment.

You do not need a client secret if you are running a client app (Web app/android app). Additionally, Cognito has historically had issues verifying the hash of client apps in the Javascript and Android AWS SDKs. If you can get by do not include it. You should not need it unless you are writing something server-side.

Current JS SDK still does not support this at the time of writing. https://github.com/aws-amplify/amplify-js/tree/master/packages/amazon-cognito-identity-js#user-content-configuration

**Comment by kbokarius** [comment]
> If you writing a server or need admin (programmatic user access without an actual user) access to Cognito ignore this comment.
> 
> You do not need a client secret if you are running a client app (Web app/android app). Additionally, Cognito has historically had issues verifying the hash of client apps in the Javascript and Android AWS SDKs. If you can get by do not include it. You should not need it unless you are writing something server-side.
> 
> Current JS SDK still does not support this at the time of writing. https://github.com/aws-amplify/amplify-js/tree/master/packages/amazon-cognito-identity-js#user-content-configuration
> https://stackoverflow.com/a/47945657/5256151

Thank you, I tried removing the client secret but we still get the `Unable to verify secret hash for client` error. Is there a setting we need to change in the Cognito client app to no longer require a client secret?

**Comment by droidkfx** [comment]
If the app you set up in AWS has a secret then you need to recreate it. You can uncheck the "generate" secret button when you configure the app.

**Comment by kbokarius** [comment]
> If the app you set up in AWS has a secret then you need to recreate it. You can uncheck the "generate" secret button when you configure the app.

Got it, I just tried creating a new app without generating the client secret. I removed the client secret from our code, replaced the old client ID with the new client ID, but the behavior is the same. Sign up works but sign in returns `Unable to verify secret hash for client`

**Comment by kbokarius** [comment]
Disregard, for some reason the code just started working after we forced a hot reload. How bizarre, our code changes seemed to be taking effect before. It looks like it's working for us now, thank you!

**Comment by droidkfx** [comment]
@kbokarius Wonderful! @furaiev There is a workaround for this issue but is there anything that can be done in the code base?

**Comment by kbokarius** [comment]
I'll let you guys close this if you don't believe this is an issue. I was definitely using the right client secret and it was failing with that error. I reviewed your code for calculating the client secret hash and it looked identical to other implementations online. Things seem to be working as expected after I removed the client secret, and as @droidkfx indicated, a client secret isn't helpful for mobile app clients.
